Add ability to parse backend blocks present in a test file's run blocks, validate configuration

SarahFrench opened the pull request:
## Description

>Note: This PR is targeted at a feature branch instead of main, as I'd like to avoid creating a single, large PR for this work.**

This PR is the first step of adding support for backend blocks to run blocks in `terraform test`:
* implement parsing of backend blocks from run blocks.
* implement validation of backend blocks in the context of runs.

Luckily we can reuse the existing code for parsing backend blocks, so validation is the main problem being addressed in this PR.

### Validation requirements

- Only state backends are supported
- Only one backend block per run block
- Only one backend block per internal state file (i.e. state_key value)
- For a given internal state file, the first command = apply run block must be where the backend block is used (if at all)

SarahFrench commented on internal/configs/test_file.go:
```diff
@@ -324,6 +326,18 @@ type TestRunOptions struct {
 
+// runBlockBackend is used when parsing a single test file as part of ensuring
+// there is only a single backend block for a given internal state file.
+// TODO(SarahFrench) this also needs to be done across multiple test files - ?
```
I need to check if parallelism work is within the scope of an individual file, or in the scope of an entire test suite.

dsa0x replied:
parallelism works within the scope of a file. in addition, runs with the same state file are never executed in parallel, and that will remain true for runs with backend

SarahFrench replied:
That's good! Identifying if the same backend is used across multiple files would be a tough thing to detect when validating configuration. If files are always sequential then we can allow users to re-use preexisting state between test files if they want/need, and if they experienced any issues it'd be simple enough for them to debug themselves.

SarahFrench commented on internal/configs/test_file.go:
```diff
@@ -648,6 +680,35 @@ func decodeTestRunBlock(block *hcl.Block, file *TestFile) (*TestRun, hcl.Diagnos
+			}
+
+			if r.Backend != nil {
+				// We've already encountered a backend for this run block
```
Another question is: do we leave it to users to make sure they avoid using the same backend configuration (i.e. the same state file) for multiple internal states within a test file?

With the introduction of parallel runs within a test file users could accidentally write tests that use the same remote state but different internal state files, which could experience conflicts in confusing ways.

dsa0x replied:
In my opinion, the state key is just a direction to the original state file, which in this case is defined by the backend configuration (and thus may be a local or remote file). Therefore, using same backend config for multiple internal states should be invalid. I cannot at the moment recall a situation where you'd want different internal states to have same state file. In that case, you can just make those runs reference the same state.

liamcervante replied:
Yes, I agree - let's prevent this if we can 👍

SarahFrench replied:
I've implemented a version of this validation in https://github.com/hashicorp/terraform/pull/36541/commits/38c284abfdf5e9d2b625ca50425a9672e74bdb78, but I'm experiencing that restriction where I cannot import code for making a hash of a backend from `internal/backend/init` into the code that parses the file in `internal/configs/`. Instead of raising diagnostics while a specific file is parsed I need to wait until the parsing completes and that data is returned to TestCommand.Run() in the `command` package.

An alternative would be to pass the necessary backend initialisation logic into the parsing logic as an argument, but that feels incorrect.

Would continuing with the approach in the commit I linked be ok?

dsa0x replied:
In addition to doing this validation in the configs package, I think another right approach for such validation would have been in the test suite itself. We already something close, where we filter the tests in the config. Doing this backend validation is however not possible at the moment due to import cycles. Given that, this approach should be sufficient for now. 

Nevertheless, the crux of the problem is because the `TestSuiteRunner` implementation is defined within the `backend` package. That was a time when tests were supposed to be limited in-memory local backend. With the advent on non-local backends, we should move it out, and thus would allow us to reference backend "parent" packages like the `init`. However, that may not need to be part of this PR
I moved that functionality into the moved test suite like I described here => https://gist.github.com/dsa0x/b076bc35f53226594496b51cc93e5670 (see `ensureBackend`)

dsa0x replied:
Sounds good 👍 I can imagine that would be beneficial beyond tests.

> Just a note that we still support Tests in TFC, but TFC ends up making a local test execution

SarahFrench replied:
I think the changes I mentioned would enable the approach outlined in [your gist](https://gist.github.com/dsa0x/b076bc35f53226594496b51cc93e5670), though I'll need to think how that would tie into other work. For now, I'm going to drop it and pick it up again if the follow on work after this PR can be improved by the refactor.

Aldocapurro reviewed (commented):
R

liamcervante reviewed (commented):
I think we're not checking anywhere that only the first run block for a given state can

dsa0x commented on internal/configs/test_file.go:
```diff
@@ -370,6 +388,50 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
 			}
 			runBlockNames[run.Name] = run.DeclRange
 
+			if run.Backend != nil {
+
+				if rb, exists := tf.BackendConfigs[run.StateKey]; exists {
```
Regarding the state key, I think what @liamcervante mentioned in an earlier comment is still relevant here.

here is an example of such implied state key from a custom module source.
```
# The implied state key for this run block is "./child"
run "test_1" {
  command = apply
  module  {
    source = "./child"
  }
}

# The implied state key for this run block is ""
# This run block does not use the same internal state as test_1,
# so it should be allowed to load its backend block
run "test_2" {
  command = apply
  backend "local" {
    path = "/tests/other-state"
  }
}
```

Here is the code that currently gathers the state key from all possibilities
https://github.com/hashicorp/terraform/blob/4161e88391f3f8f4ded60ec0e61a8712d09bdced/internal/moduletest/run.go#L191-L202

Perhaps that could be done within the configs package instead, i.e setting the run's state key as the module source.

SarahFrench replied:
Yep, working on this :)

dsa0x commented on internal/configs/testdata/valid-modules/with-tests-backend/test_case_two.tftest.hcl:
```diff
@@ -0,0 +1,15 @@
+# This backend is used to set the internal state "foobar-1"
+run "test_1" {
+  state_key = "foobar-1"
+  backend "local" {
+    path = "state/terraform.tfstate"
+  }
+}
+
+# This backend is used to set the internal state "foobar-2"
+run "test_2" {
+  state_key = "foobar-2"
+  backend "local" {
+    path = "state/terraform.tfstate"
+  }
+}
```
Should this not be invalid? We have 2 internal states working on the same backend.

SarahFrench replied:
Indeed, that's now invalid. But the test hasn't failed because the validation is currently [38c284a](https://github.com/hashicorp/terraform/pull/36541/commits/38c284abfdf5e9d2b625ca50425a9672e74bdb78) implemented at a higher level and isn't visible to this test. This is one of my motivations for [seeking feedback here](https://github.com/hashicorp/terraform/pull/36541#discussion_r1975442336), as splitting the validation to different places causes confusion like this.

liamcervante commented on internal/configs/test_file.go:
```diff
@@ -701,12 +794,36 @@ func decodeTestRunBlock(block *hcl.Block, file *TestFile) (*TestRun, hcl.Diagnos
+		} else {
+			// Otherwise, an empty string is returned, and this denotes that the
+			// run is using the root module under test.
+			r.StateKey = ""
```
nit: this is redundant, the state key is already empty

SarahFrench replied:
Thanks! 🤦🏻  I've left the comment in place but removed the unnecessary assignment: [54dd772](https://github.com/hashicorp/terraform/pull/36541/commits/54dd772104a61c2500b8826e223ceed7cadd1ad7)

SarahFrench replied:
We could potentially change the state key for the root module to stop being an empty string, but I imagine that would add more validation issues (i.e stop name collision with a user-set state key)

SarahFrench commented on internal/command/test_test.go:
```diff
@@ -2564,6 +2564,52 @@ Failure! 0 passed, 1 failed.
+// TestTest_ReusedBackendConfiguration asserts that it's not valid to re-use the same backend config (i.e the same state file)
+// in parallel runs. This would result in multiple actions attempting to set state, potentially with different resource configurations.
+func TestTest_ReusedBackendConfiguration(t *testing.T) {
+	t.Skip("TODO(SarahFrench): Unable to reproduce this flakiness locally, skipping for now. Need to investigate or remove.") // Similar failure to TestTest_InvalidConfig?
```
It's a different issue - [the test failure](https://github.com/hashicorp/terraform/actions/runs/13632143097/job/38102077385#step:6:39) is due to the non-deterministic order that runs are processed in if they have different state keys. I believe the graph can process either of them first, so they're effectively racing to be the first run that's registered as using that backend config.

I've solved it by making the test assert the error message content but being looser about asserting which block at which line number is at fault: [a84553d](https://github.com/hashicorp/terraform/pull/36541/commits/a84553df35fce541e43458eb3fc870e5660f7786)

dsa0x replied:
It seems to me that the error we are checking here occurs in `internal/command/test.go`. At that point, the graph is not involved at all. I think that the reason for this non-determinism is because `BackendConfig` is a map. 

Is it fine for the test error to be non-deterministic? We could decide that it's not, and simply sort BackendConfig before checking the duplicate validation

dsa0x replied:
Btw, If we decide to make deterministic, here is an example in https://pkg.go.dev/iter#hdr-Standard_Library_Usage uses the new iterator packages to sort map keys.

SarahFrench replied:
(Marking this unresolved again)

I took a look - that's a neat trick for sorting map keys! For the code I have the map goes from state_key to backend config, so map keys aren't what's needed for sorting. In [75dfe43](https://github.com/hashicorp/terraform/pull/36541/commits/75dfe43978b2ab7cdbe5c9959184d9002540aab0) I've added sorting by the position the backend block is declared in the file.

SarahFrench commented on internal/configs/test_file.go:
```diff
@@ -370,6 +384,23 @@ func loadTestFile(body hcl.Body) (*TestFile, hcl.Diagnostics) {
 			}
 			runBlockNames[run.Name] = run.DeclRange
 
+			if rb, exists := stateKeyBackend[run.StateKey]; exists && run.Backend != nil {
```
Addressed in [158151c](https://github.com/hashicorp/terraform/pull/36541/commits/158151cd657a65ebebf13ca48e998546fc7bc225)

SarahFrench commented:
The crashes resulting from [feb5a7b](https://github.com/hashicorp/terraform/pull/36541/commits/feb5a7b53e91008250da177e85e8de468da66f36) show that we're not setting the state key properly after the changes in [158151c](https://github.com/hashicorp/terraform/pull/36541/commits/158151cd657a65ebebf13ca48e998546fc7bc225).

I think that's because we don't set ConfigUnderTest until here: https://github.com/hashicorp/terraform/blob/3f4ed7e01921e7b9e8943875cf8bc97a3751395f/internal/configs/config_build.go#L34

SarahFrench commented:
The latest changes have:
* Made sure that state keys for child modules are set during parsing of test files
* Removed use of GetStateKey
* Added some more tests that can adapt to runs being processed in different orders

SarahFrench commented:
Sorry for the back and forth here! Previously I thought the different ordering that impacted the acceptance test were due to new behaviour in the test graph, but it was just due to iterating over a map 🤦🏻 . I've addressed that in [75dfe43](https://github.com/hashicorp/terraform/pull/36541/commits/75dfe43978b2ab7cdbe5c9959184d9002540aab0)

dsa0x reviewed (approved):
Thanks for addressing that last bit 🫡

SarahFrench commented:
Thanks both! 🙇🏻
